google/cloud/storage/internal/curl_download_request.cc

```diff
@@ -28,76 +28,141 @@ namespace storage {
 inline namespace STORAGE_CLIENT_NS {
 namespace internal {
 
-CurlDownloadRequest::CurlDownloadRequest(std::size_t initial_buffer_size)
+// Note that TRACE-level messages are disabled by default, even in
+// CMAKE_BUILD_TYPE=Debug builds. The level of detail created by the
+// TRACE_STATE() macro is only needed by the library developers when
+// troubleshooting this class.
+#define TRACE_STATE()                                                       \
+  GCP_LOG(TRACE) << __func__ << "(), buffer_size_=" << buffer_size_         \
+                 << ", buffer_offset_=" << buffer_offset_                   \
+                 << ", spill_.size()=" << spill_.size()                     \
+                 << ", spill_offset_=" << spill_offset_                     \
+                 << ", closing=" << closing_ << ", closed=" << curl_closed_ \
+                 << ", paused=" << paused_ << ", in_multi=" << in_multi_
+
+CurlDownloadRequest::CurlDownloadRequest()
     : headers_(nullptr, &curl_slist_free_all),
       multi_(nullptr, &curl_multi_cleanup),
-      initial_buffer_size_(initial_buffer_size) {
-  buffer_.reserve(initial_buffer_size);
+      spill_(CURL_MAX_WRITE_SIZE) {}
+
+template <typename Predicate>
+Status CurlDownloadRequest::Wait(Predicate predicate) {
+  int repeats = 0;
+  // We can assert that the current thread is the leader, because the
+  // predicate is satisfied, and the condition variable exited. Therefore,
+  // this thread must run the I/O event loop.
+  while (!predicate()) {
+    handle_.FlushDebug(__func__);
+    TRACE_STATE() << ", repeats=" << repeats;
+    auto running_handles = PerformWork();
+    if (!running_handles.ok()) {
+      return std::move(running_handles).status();
+    }
+    // Only wait if there are CURL handles with pending work *and* the
+    // predicate is not satisfied. Note that if the predicate is ill-defined
+    // it might continue to be unsatisfied even though the handles have
+    // completed their work.
+    if (*running_handles == 0 || predicate()) {
+      return Status();
+    }
+    auto status = WaitForHandles(repeats);
```

Comment:
Nit: It is slightly confusing that the argument is in/out, but that's just an opinion.

Reply:
I somewhat agree, but it is a private function of a class in the internal namespace, I going to leave as-is.